Mta 5505 (#4344)
scaleway/docs-content: pages/key-manager/api-cli/encrypt-decrypt-keys-with-streaming-aead-tink.mdx

Tink is a multi-language cryptographic library that simplifies common cryptographic operations like encryption, decryption, signing, and more. It provides cryptographic primitives, including AEAD.

Key Manager handles secure storage and management of encryption keys. In this case, it is integrated with Tink to manage a Key Encryption Key (KEK).

Streaming AEAD: This is the cryptographic primitive that handles secure, authenticated streaming encryption. It ensures that large amounts of data are encrypted on the fly without needing to load everything into memory.

This page shows you how to use the Key Manager Tink extension to encrypt and decrypt large data streams using Authenticated Encryption with Associated Data (AEAD).

<Macro id="requirements" />

- A Scaleway account logged into the [console](https://console.scaleway.com)
- [Owner](/iam/concepts/#owner) status or [IAM permissions](/iam/concepts/#permission) allowing you to perform actions in the intended Organization
- Created a key encryption key (KEK) in Key Manager
- A working Go environment
- Installed the [Scaleway Go SDK](https://github.com/scaleway/scaleway-sdk-go) and created [credentials](/scaleway-sdk/go-sdk/)

> - [Added the necessary dependencies](/key-manager/api-cli/configuring-tink/#setting-up-tink) for your Go project

Reviewer: Project or project?

        c, err := os.ReadFile(ciphertextPath)
        if err != nil {
            log.Fatal(err)
        }
        fmt.Println(string(c))

        // Decrypt the ciphertext file and write the output in the decrypted file.
        ciphertextFile, err = os.Open(ciphertextPath)
        if err != nil {
            log.Fatal(err)
        }
        decryptedPath := filepath.Join(dir, "decrypted")
        decryptedFile, err := os.Create(decryptedPath)
        if err != nil {
            log.Fatal(err)
        }
        r, err := primitive.NewDecryptingReader(ciphertextFile, associatedData)
        if err != nil {
            log.Fatal(err)
        }
        if _, err := io.Copy(decryptedFile, r); err != nil {
            log.Fatal(err)
        }
        if err := decryptedFile.Close(); err != nil {
            log.Fatal(err)
        }
        if err := ciphertextFile.Close(); err != nil {
            log.Fatal(err)
        }

        // Print the content of the decrypted file.
        b, err := os.ReadFile(decryptedPath)
        if err != nil {
            log.Fatal(err)
        }
        fmt.Println(string(b))
    }
    ```
    <Message type="note">

>      - The example above shows you how to use the Key Manager remote key encryption key to protect your data encryption key and AEAD stream, you can also protect the data encryption key using another non-remote key.

Reviewer: I'm not sure that I understood this sentence. There are a lot of mentions to 'key', so I keep restarting and not understanding. But maybe it's just me. 😅

>      - The example above shows you how to use the Key Manager remote key encryption key to protect your data encryption key and AEAD stream, you can also protect the data encryption key using another non-remote key.

Author: ```suggestion
     - The example above demonstrates how to use the Key Manager's remote encryption key to protect your data encryption key and AEAD stream. Alternatively, you can use a non-remote key to protect the data encryption key instead.
```
